### PR TITLE
Fix PdsRequired for local CPhysicalAgg with DQAs

### PR DESCRIPTION
**Pull request by zellerh**
When we had a local aggregate with a DQA (e.g. select count(distinct b)
... group by a), we generated a required distribution on the DQA column
(b in this example). We now also consider an alternative distribution
spec with both the groupby and the DQA columns (a, b in the example).

Note: We have some MDP tests for multiple DQAs which are not supported
in ORCA - when run through PSQL the same query will fall back. We
don't generate good plans for those and we should probably delete
those MDPs (MDQA*.mdp).

Co-authored-by: Sambitesh Dash <sdash@pivotal.io>

### CODE REVIEW

File: libgpopt/src/operators/CPhysicalAgg.cpp [hunk 1]
```diff
@@ -86,9 +86,13 @@ CPhysicalAgg::CPhysicalAgg
...
 		{
 			// If the local aggregate has distinct columns we generate
-			// one optimization requests for its children:
-			// (1) hash distribution on the distinct columns
-			ulDistrReqs = 1;
+			// two optimization requests for its children:
+			// (1) hash distribution on the distinct columns only
+			// (2) hash distribution on the grouping and distinct
+			//     columns (only if the grouping columns are not empty)
+			if (0 == m_pdrgpcr->Size()) {
```
Comment (bhuvnesh2703) [inline libgpopt/src/operators/CPhysicalAgg.cpp]: is this array always initialized? do we need to check it for `!= NULL` too?

File: libgpopt/src/operators/CPhysicalAgg.cpp [hunk 2]
```diff
@@ -235,8 +239,19 @@ CPhysicalAgg::PdsRequiredAgg
 
 	if (COperator::EgbaggtypeLocal == m_egbaggtype && m_pdrgpcrArgDQA != NULL && 0 != m_pdrgpcrArgDQA->Size())
 	{
-		GPOS_ASSERT(0 == ulOptReq);
-		return PdsMaximalHashed(mp, m_pdrgpcrArgDQA);
+		if (ulOptReq == 0) {
```
Comment (bhuvnesh2703) [inline libgpopt/src/operators/CPhysicalAgg.cpp]: very minor, `{` on next line..

File: libgpopt/src/operators/CPhysicalAgg.cpp [hunk 3]
```diff
@@ -235,8 +239,19 @@ CPhysicalAgg::PdsRequiredAgg
...
+			return PdsMaximalHashed(mp, m_pdrgpcrArgDQA);
+		}
+		else
+		{
+			GPOS_ASSERT(1 == ulOptReq);
+			CColRefArray *combinedCols = GPOS_NEW(mp) CColRefArray(mp);
```
Comment (bhuvnesh2703) [inline libgpopt/src/operators/CPhysicalAgg.cpp]: var name may be .. grp_and_distinct_cols

File: libgpopt/src/operators/CPhysicalAgg.cpp [hunk 4]
```diff
@@ -235,8 +239,19 @@ CPhysicalAgg::PdsRequiredAgg
...
+			combinedCols->AppendArray(m_pdrgpcr);
```
Comment (bhuvnesh2703) [inline libgpopt/src/operators/CPhysicalAgg.cpp]: should we assert m_pdrgpcr != NULL

File: libgpopt/src/operators/CPhysicalAgg.cpp (hunk 2 above)
Comment (zellerh) [reply libgpopt/src/operators/CPhysicalAgg.cpp]: Done.

File: libgpopt/src/operators/CPhysicalAgg.cpp (hunk 1 above)
Comment (zellerh) [reply libgpopt/src/operators/CPhysicalAgg.cpp]: I checked and all the paths that call the CPhysicalAgg constructor seem to call AddRef() on the ColRefArray that is passed in for m_pdrgpcr, so I think it's safe to assume that it is non-null. There also is already an assert in the constructor that ensures the pointer is not NULL.

File: libgpopt/src/operators/CPhysicalAgg.cpp (hunk 3 above)
Comment (zellerh) [reply libgpopt/src/operators/CPhysicalAgg.cpp]: Done.

File: libgpopt/src/operators/CPhysicalAgg.cpp (hunk 4 above)
Comment (zellerh) [reply libgpopt/src/operators/CPhysicalAgg.cpp]: That assert already exists, in the CPhysicalAgg constructor.

**Pull request merged**